elements/lisk-p2p/src/utils.ts

```diff
@@ -0,0 +1,225 @@
+	if (network !== NETWORK.NET_IPV4) {
+		return cryptoHashSha256(
+			Buffer.concat([secretBytes, networkBytes]),
+		).readUInt32BE(0);
+	}
+
+	const netgroupBytes = Buffer.concat([
+		secretBytes,
+		networkBytes,
+		aBytes,
+		bBytes,
+	]);
+
+	return cryptoHashSha256(netgroupBytes).readUInt32BE(0);
+};
+
+// For new peer buckets, provide the source IP address from which the peer list was received
+export const getBucket = (options: {
+	readonly secret: number;
+	readonly targetAddress: string;
+	readonly sourceAddress?: string;
+}): number => {
+	const { secret, targetAddress, sourceAddress } = options;
+	const peerListType = sourceAddress ? NEW_PEERS : TRIED_PEERS;
+	const firstMod = peerListType === NEW_PEERS ? BYTES_16 : BYTES_4;
+	const secondMod = peerListType === NEW_PEERS ? BYTES_128 : BYTES_64;
+	const secretBytes = Buffer.alloc(SECRET_BUFFER_LENGTH);
+	secretBytes.writeUInt32BE(secret, 0);
+	const network = getNetwork(targetAddress);
+	const networkBytes = Buffer.alloc(NETWORK_BUFFER_LENGTH);
+	networkBytes.writeUInt8(network, 0);
+
+	// Get bytes of ip address to bucket
+	const {
+		aBytes: targetABytes,
+		bBytes: targetBBytes,
+		cBytes: targetCBytes,
+		dBytes: targetDBytes,
+	} = getIPBytes(targetAddress);
+
+	// Get prefix bytes of source address from which peer list is received
+	const { aBytes: sourceABytes, bBytes: sourceBBytes } = getIPBytes(
+		targetAddress,
+	);
+
+	// Check if ip address is unsupported network type
+	if (network === NETWORK.NET_OTHER) {
+		throw Error('IP address is unsupported.');
+	}
+
+	// Seperate buckets for local and private addresses
+	if (network !== NETWORK.NET_IPV4) {
+		return (
+			cryptoHashSha256(Buffer.concat([secretBytes, networkBytes])).readUInt32BE(
+				0,
+			) % secondMod
+		);
+	}
+
+	const addressBytes = Buffer.concat([
+		targetABytes,
+		targetBBytes,
+		targetCBytes,
+		targetDBytes,
+	]);
+
+	// New peers: k = Hash(random_secret, source_group, group) % 16
+	// Tried peers: k = Hash(random_secret, IP) % 4
+	const kBytes = Buffer.alloc(firstMod);
+	const k =
+		peerListType === NEW_PEERS
+			? cryptoHashSha256(
+					Buffer.concat([
+						secretBytes,
+						networkBytes,
+						sourceABytes,
+						sourceBBytes,
+						targetABytes,
+						targetBBytes,
+					]),
+			  ).readUInt32BE(0) % firstMod
+			: cryptoHashSha256(
+					Buffer.concat([secretBytes, networkBytes, addressBytes]),
+			  ).readUInt32BE(0) % firstMod;
+
+	kBytes.writeUInt32BE(k, 0);
+
+	// New peers: b = Hash(random_secret, source_group, k) % 128
+	// Tried peers: b = Hash(random_secret, group, k) % 64
+	const bucketBytes =
+		peerListType === NEW_PEERS
+			? Buffer.concat([
+					secretBytes,
+					networkBytes,
+					sourceABytes,
+					sourceBBytes,
+					kBytes,
+			  ])
+			: Buffer.concat([
+					secretBytes,
+					networkBytes,
+					targetABytes,
+					targetBBytes,
+					kBytes,
+			  ]);
+
+	return cryptoHashSha256(bucketBytes).readUInt32BE(0) % secondMod;
```

Comment:
in general, do we need to use `0` offset?

Reply:
you don't need it to work but linter complains without it